File: Tester/BookmarksSample.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using FastColoredTextBoxNS;
using System.Drawing.Drawing2D;

namespace Tester
{
    public partial class BookmarksSample : Form
    {
        // Key - Line.UniqueId
        HashSet<int> bookmarksLineId = new HashSet<int>();
        // Index - bookmark number, Value - Line.UniqueId
        List<int> bookmarks = new List<int>();

        public BookmarksSample()
        {
            InitializeComponent();
        }

        private void fctb_PaintLine(object sender, PaintLineEventArgs e)
        {
            //draw bookmark
            if (bookmarksLineId.Contains(fctb[e.LineIndex].UniqueId))
            {
                e.Graphics.FillEllipse(new LinearGradientBrush(new Rectangle(0, e.LineRect.Top, 15, 15), Color.White, Color.PowderBlue, 45), 0, e.LineRect.Top, 15, 15);
                e.Graphics.DrawEllipse(Pens.PowderBlue, 0, e.LineRect.Top, 15, 15);
            }
        }

        private void btAddBookmark_Click(object sender, EventArgs e)
        {
            //get UniqueId of current line
            int id = fctb[fctb.Selection.Start.iLine].UniqueId;
            if (bookmarksLineId.Contains(id))
                return;
            //add bookmark
            bookmarks.Add(id);
            bookmarksLineId.Add(id);
            //repaint
            fctb.Invalidate();
        }

        private void btRemoveBookmark_Click(object sender, EventArgs e)
        {
            //get UniqueId of current line
            int id = fctb[fctb.Selection.Start.iLine].UniqueId;
            if (!bookmarksLineId.Contains(id))
                return;
            //remove bookmark
            bookmarks.Remove(id);
            bookmarksLineId.Remove(id);
            //repaint
            fctb.Invalidate();
        }

        private void fctb_LineRemoved(object sender, LineRemovedEventArgs e)
        {
            //remove lines from bookmarks
            foreach (int id in e.RemovedLineUniqueIds)
            if(bookmarksLineId.Contains(id))
            {
                bookmarksLineId.Remove(id);
                bookmarks.Remove(id);
            }
        }

        private void btGo_DropDownOpening(object sender, EventArgs e)
        {
            btGo.DropDownItems.Clear();
            for (int i = 0; i < bookmarks.Count; i++)
            {
                var item = btGo.DropDownItems.Add("Bookmark " + i);
                item.Tag = i;
                item.Click += (o, a) => GoTo((int)(o as ToolStripItem).Tag);
            }
        }

        private void GoTo(int iBookmark)
        {
            if (iBookmark < 0 || iBookmark >= bookmarks.Count)
                return;

            int id = bookmarks[iBookmark];
            for(int i=0;i<fctb.LinesCount;i++)
                if(fctb[i].UniqueId == id)
                {
                    fctb.Selection.Start = new Place(0, i);
                    fctb.DoSelectionVisible();
                    fctb.Invalidate();
                    break;
                }
        }
    }
}

File: TesterVB/BookmarksSample.vb
Imports System.Drawing.Drawing2D

Public Class BookmarksSample
    ' Key - Line.UniqueId
    Private bookmarksLineId As New Dictionary(Of Integer, Integer)()
    ' Index - bookmark number, Value - Line.UniqueId
    Private bookmarks As New List(Of Integer)()

    Private Sub fctb_LineRemoved(ByVal sender As System.Object, ByVal e As FastColoredTextBoxNS.LineRemovedEventArgs) Handles fctb.LineRemoved
        'remove lines from bookmarks
        For Each id As Integer In e.RemovedLineUniqueIds
            If bookmarksLineId.ContainsKey(id) Then
                bookmarksLineId.Remove(id)
                bookmarks.Remove(id)
            End If
        Next
    End Sub


    Private Sub fctb_PaintLine(ByVal sender As System.Object, ByVal e As FastColoredTextBoxNS.PaintLineEventArgs) Handles fctb.PaintLine
        'draw bookmark
        If bookmarksLineId.ContainsKey(fctb(e.LineIndex).UniqueId) Then
            e.Graphics.FillEllipse(New LinearGradientBrush(New Rectangle(0, e.LineRect.Top, 15, 15), Color.White, Color.PowderBlue, 45), 0, e.LineRect.Top, 15, 15)
            e.Graphics.DrawEllipse(Pens.PowderBlue, 0, e.LineRect.Top, 15, 15)
        End If
    End Sub

    Private Sub GoToBookmark(ByVal iBookmark As Integer)
        If iBookmark < 0 OrElse iBookmark >= bookmarks.Count Then
            Return
        End If

        Dim id As Integer = bookmarks(iBookmark)
        For i As Integer = 0 To fctb.LinesCount - 1
            If fctb(i).UniqueId = id Then
                fctb.Selection.Start = New Place(0, i)
                fctb.DoSelectionVisible()
                fctb.Invalidate()
                Exit For
            End If
        Next
    End Sub

    Private Sub btGo_DropDownOpening(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btGo.DropDownOpening
        btGo.DropDownItems.Clear()
        For i As Integer = 0 To bookmarks.Count - 1
            Dim item = btGo.DropDownItems.Add("Bookmark " & i)
            item.Tag = i
        Next
    End Sub

    Private Sub btRemoveBookmark_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btRemoveBookmark.Click
        'get UniqueId of current line
        Dim id As Integer = fctb(fctb.Selection.Start.iLine).UniqueId
        If Not bookmarksLineId.ContainsKey(id) Then
            Return
        End If
        'remove bookmark
        bookmarks.Remove(id)
        bookmarksLineId.Remove(id)
        'repaint
        fctb.Invalidate()
    End Sub

    Private Sub btAddBookmark_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btAddBookmark.Click
        'get UniqueId of current line
        Dim id As Integer = fctb(fctb.Selection.Start.iLine).UniqueId
        If bookmarksLineId.ContainsKey(id) Then
            Return
        End If
        'add bookmark
        bookmarks.Add(id)
        bookmarksLineId.Add(id, id)
        'repaint
        fctb.Invalidate()
    End Sub

    Private Sub btGo_DropDownItemClicked(ByVal sender As System.Object, ByVal e As System.Windows.Forms.ToolStripItemClickedEventArgs) Handles btGo.DropDownItemClicked
        GoToBookmark(CInt(e.ClickedItem.Tag))
    End Sub
End Class